script/model/data_class/opunit_data.py

```diff
@@ -50,40 +50,79 @@ def _default_get_mini_runner_data(filename):
-            if opunit not in data_info.ARITHMETIC_OPUNITS or line_data[arithmetic_mode_index] != '1':
-                data_map[opunit].append(list(map(int, line_data)))
+            # drop query_id, pipeline_id, num_features, features_vector
+            record = [d for i,d in enumerate(line) if i > features_vector_index]
+            record.insert(data_info.EXECUTION_MODE_INDEX, line[execution_mode_index])
+            data = list(map(_convert_string_to_numeric, record))
+            x_multiple = data[:data_info.RECORD_FEATURES_END]
+            y_merged = np.array(data[-data_info.RECORD_METRICS_START:])
+
+            # Get the opunits located within
+            opunits = []
+            features = line[features_vector_index].split(';')
+            for idx, feature in enumerate(features):
+                opunit = OpUnit[feature]
+                if opunit not in data_info.ARITHMETIC_OPUNITS or line[execution_mode_index] != '1':
+                    x_loc = [v[idx] if type(v) == list else v for v in x_multiple]
+                    if opunit in model_map:
+                        key = [opunit] + x_loc
+                        if tuple(key) in predict_cache:
+                            y_merged = y_merged - predict_cache[tuple(key)]
+                        else:
+                            predict = model_map[opunit].predict(np.array(x_loc).reshape(1, -1))[0]
+                            predict_cache[tuple(key)] = predict
+                            y_merged = y_merged - predict
+
+                        y_merged = np.clip(y_merged, 0, None)
+                    else:
+                        opunits.append((opunit, x_loc))
+
+            if len(opunits) > 1:
+                raise Exception('Unmodelled OperatingUnits detected: {}'.format(opunits))
+
+            # record real result
+            predict_cache[tuple([opunits[0][0]] + opunits[0][1])] = list(y_merged)
+
+            # opunits[0][0] is the opunit
+            # opunits[0][1] is input feature
+            # y_merged should be post-subtraction
+            if opunits[0][0] not in data_map:
+                data_map[opunits[0][0]] = []
+            data_map[opunits[0][0]].append(opunits[0][1] + list(y_merged))
 
     data_list = []
-    # Since the compiled arithmetics are always optimized away,
-    # we treat the compiled arithmetics the same as the interpreted ones by copying the data
     for opunit, values in data_map.items():
-        print(opunit)
+        # Since the compiled arithmetics are always optimized away,
+        # we treat the compiled arithmetics the same as the interpreted ones by copying the data
         if opunit in data_info.ARITHMETIC_OPUNITS:
             compiled_values = copy.deepcopy(values)
             for v in compiled_values:
-                v[arithmetic_mode_index] = 1
+                v[data_info.EXECUTION_MODE_INDEX] = 1 # mark as compiled
             values += compiled_values
```

Comment:
I think we can remove this code block now. Since we can control what features to generate in the mini-runner, we can run the arithmetics for the second time and mark them as the "compiled" mode. Then the code would be more consistent.